### PR TITLE
Introduce DisplayRow, MultiBufferRow newtypes and BufferRow type alias

### PR DESCRIPTION
**Pull request by SomeoneToIgnore**
Part of https://github.com/zed-industries/zed/issues/8081

To avoid confusion and bugs when converting between various row `u32`'s, use different types for each.
Further PRs should split `Point` into buffer and multi buffer variants and make the code more readable.

Release Notes:

- N/A

### CODE REVIEW

File: crates/editor/src/display_map/block_map.rs [hunk 1]
```diff
@@ -700,9 +700,9 @@ impl BlockSnapshot {
         }
     }
 
-    pub fn is_block_line(&self, row: u32) -> bool {
+    pub fn is_block_line(&self, row: DisplayRow) -> bool {
```
Comment (osiewicz) [inline crates/editor/src/display_map/block_map.rs:703]: Same question here; should block_map api be defined in terms of `BlockRow`s?

File: crates/editor/src/display_map/block_map.rs [hunk 2]
```diff
@@ -885,12 +885,12 @@ impl<'a> Iterator for BlockBufferRows<'a> {
 
     fn next(&mut self) -> Option<Self::Item> {
```
Comment (osiewicz) [inline crates/editor/src/display_map/block_map.rs:886]: In the same vein, should `BlockBufferRows::Item` be a `Option<BlockRow>` and not `Option<u32>`?

File: crates/editor/src/display_map/inlay_map.rs [hunk 3]
```diff
@@ -375,17 +377,17 @@ impl<'a> InlayBufferRows<'a> {
         self.transforms.seek(&inlay_point, Bias::Left, &());
 
         let mut buffer_point = self.transforms.start().1;
-        let buffer_row = if row == 0 {
+        let buffer_row = MultiBufferRow(if row == 0 {
             0
         } else {
             match self.transforms.item() {
                 Some(Transform::Isomorphic(_)) => {
                     buffer_point += inlay_point.0 - self.transforms.start().0 .0;
```
Comment (osiewicz) [inline crates/editor/src/display_map/inlay_map.rs:385]: This is kind of suspicious tbh

Comment (SomeoneToIgnore) [reply crates/editor/src/display_map/inlay_map.rs:385]: Can you elaborate?

File: crates/editor/src/editor.rs [hunk 4]
```diff
@@ -2498,7 +2498,8 @@ impl Editor {
         let end_column = cmp::max(tail.column(), goal_column);
         let reversed = start_column < tail.column();
 
-        let selection_ranges = (start_row..=end_row)
+        let selection_ranges = (start_row.0..=end_row.0)
```
[the remark this answers is not in the document]
Comment (SomeoneToIgnore) [reply crates/editor/src/editor.rs:2501]: Added.

File: crates/editor/src/display_map/inlay_map.rs (hunk 3 above)
Comment (osiewicz) [reply crates/editor/src/display_map/inlay_map.rs:385]: I mean that it seems a bit suspicious how we intermix inlay points and multibuffer points ig

Comment (SomeoneToIgnore) [reply crates/editor/src/display_map/inlay_map.rs:385]: That code seems fine, just quite specific: it tries to find an isomorphic transform boundary (in our case, end of the buffer text just before we insert the inlay hint), and uses the end of that buffer text as a start of the inlay row.

@\as-cii was the one who wrote it and knows it better, I trust him and our fuzzy tests that this works 🙂

**Pull request merged**